### PR TITLE
feat: Adding self-extracting file as an optional feature

### LINKED ISSUE #4: Self-extracting files
It would be interesting to explore self-extracting files as an optional feature similar to how conda-constructor does it.

### PR DESCRIPTION
**Pull request by prsabahrami**
# Motivation
Closes #4 
# Changes
Similar to `conda-constructor`, I have added a header.sh file. After the environment has been packed and archived, that archive is appended onto header.sh which later extracts it when it runs.

### Packing an environment
```bash
pixi-pack pack --create-executable
```
### Unpacking an environment
either directly run the .sh file created
```bash
sh environment.sh
```
You can use below flags
```txt
Unpacks an environment packed with pixi-pack

-f           no error if environment already exists
-h           print this help message and exit
-p ENV       environment prefix, defaults to $PREFIX
-i INSTALLER create the environment using the specified installer defaulting to $INSTALLER
-a           create an activation script to activate the environment
```
or you can directly use
```bash
pixi-pack unpack environment.sh
```
I will try to add better tests and error handling in the upcoming days. Meanwhile, let me know if you have any recommendations.

### CODE REVIEW

**Comment by 0xbe7a** [comment]
One of the biggest use-cases for us for pixi-pack is the ability to build up archives on Linux for foreign targets, like Windows and to extract it there. Having a self-extracting archive there would be probably the most useful scenario. Your prepended shell script will only work for Unix systems, right?

**Comment by prsabahrami** [comment]
@0xbe7a 
Yes, it is correct. 
I was thinking I could add a `bat` self-extracting script to be used on Windows in a separate PR. Then through the passed `--platform` we can decide to generate `sh` script or `bat` script (or both).
Please let me know if this works for the mentioned scenario.

Also, in regards to the extraction tool used in the script, fair enough. 
I was thinking to implement the logic used in `unpack.rs` onto the script(s) so we could get rid of `conda` and `micromambda`. Let me know what you think. I'm not sure how efficient the script will turn out.

**Comment by 0xbe7a** [comment]
The part that involves extracting and linking the packages is highly non-trivial and probably not really feasible with a bash script. Instead, we should try using `rattler` as we did in `pixi-pack` itself. `rattler` supports all the platforms we target. The approach I am considering involves including an executable object with the required unpack functionality and using this when unpacking the archive to make it truly standalone.

Additionally, upon reviewing your pull request again, please note that `micromamba shell activate` already exists and can create the entire activation script for `micromamba`. For unpacking the archive using micromamba, you can use micromamba to create the environment and then generate the activation script to get the final prefix. This way, a single shell script can be used to activate into this environment, similar to what pixi-unpack would have produced.

**Comment by prsabahrami** [comment]
Sounds good. 
I will include an executable using `rattler` onto the script. Which the script will use to properly extract the archive. 

Also, in regards to the `micromamba shell activate`, I will try to use `rattler_shell` inside the executable to create the activation script so we can remove the necessitiy of `conda` or `micromambda`.

**Comment by prsabahrami** [comment]
Apologies for the delay. Got sucked up into school work.
I have added an `extractor` subproject which then is built and its binary is added to script. 
The original commands still work but in case the `-i conda` is passed to the script, one can not create an activation script since I'm using `micromamba` or `rattler` to create the script. If needed, I can modify header.sh to extract `extractor` binary at all times but I thought if someone has conda installed on their devices, there is no need to take extra time and extract the binary. 
Let me know what you think and then I can improve the verbosity and testing.
**Note:** The logic used in extractor is just a simpler version in unpack.rs.

**Comment by pavelzw** [comment]
> there is no need to take extra time and extract the binary

I think it would be better for reproducibility if we always unpack the same way, i.e. always use `extractor`

**Comment by pavelzw** [comment]
Could you fix the merge conflicts as well?

**Comment by prsabahrami** [comment]
> Could you fix the merge conflicts as well?

Done.

**Comment by prsabahrami** [comment]
I believe the issue with the lockfile should be fixed now.

**Comment by prsabahrami** [comment]
Added the tests from `unpack.rs` to `extractor`.
One thing I don't like is that the dependencies of `extractor` are being defined separately and hence built and compiled twice during the build process. Usually, using `[workspace]` solves this in `Cargo.toml` but I couldn't make it work using `[workspace]`. Any suggestions? This is making the build somewhat slow.

**Comment by prsabahrami** [comment]
Fixed the merge conflict. 
Any idea on the previous comment?

**Comment by prsabahrami** [comment]
As I mentioned above, the issue here is with the lockfile conflicts and incorrect dependencies between `extractor` and `pixi-pack`. I am not sure how to correctly take care of this but for now, I have updated the dependencies and made `extractor` compliant with `rattler`'s latest version. But once again, if the lockfile of the main project changes, I am not sure how it will affect `extractor`.

File: extractor/Cargo.toml
Comment (pavelzw) [inline extractor/Cargo.toml:1]: I think the extractor with the additional cargo project is too complex. How about the following setup:

`pixi-pack pack --executable` (not sure about `--executable` but something along the lines) downloads the `pixi-pack` binary for the corresponding platform from `https://github.com/Quantco/pixi-pack/releases/download/<version-of-pixi-pack>/pixi-pack-aarch64-unknown-linux-musl`, puts this into the actual executable (including `header.sh` which calls the appropriate `pixi-pack unpack` command)

This way we don't have the issue of drifting dependencies between the projects.
This _should_ be okay from a file size perspective

File: src/header.sh [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+while getopts ":fhai:p:" x; do
+    case "$x" in
+        f)
+            FORCE=1
+            ;;
+        p)
+            PREFIX="$OPTARG"
+            ;;
+        i)
+            INSTALLER="$OPTARG"
```
Comment (pavelzw) [inline src/header.sh:37]: i don't think we need different installers, let's only support our own pixi-pack one that we directly ship in the binary

File: src/header.sh
Comment (pavelzw) [inline src/header.sh:1]: `header.sh` won't work on Windows.
I would strongly prefer if we supported Windows cross-platform executable packs as well as this would be the IMO most relevant use case.

For this, we would need to read up a bit on `.exe` bundling, though.

**Review by pavelzw — CHANGES_REQUESTED** [review]

File: extractor/Cargo.toml
Comment (prsabahrami) [reply extractor/Cargo.toml:1]: I like this. I'll start working on it! 
The only difference would be size of final header.sh which is not much of a difference from extractor.

File: src/header.sh
Comment (prsabahrami) [reply src/header.sh:1]: Absolutely! 
I'll work on adding a `header.exe` or something similar.

File: extractor/Cargo.toml
Comment (prsabahrami) [reply extractor/Cargo.toml:1]: @pavelzw In order to do this, we need to know find out if the linux target os is compiled against musl or GNU libc. That is not something `rattler_conda_types` provides with the `Platform` enum. Do you think in linux case, we should add two executables and see which works? =) That is somewhat odd. The other option would be to default to `gnu` since it is more common, and modify if given as input. IINW, Alpine distributions are compiled against musl, right?

Comment (pavelzw) [reply extractor/Cargo.toml:1]: You can also execute musl binaries on systems that use gnu glibc 
It only doesn't work the other way around. That's why I would just download the musl variant (the gnu variant requires OpenSSL to be installed iirc)

Side note: Technically, the target system should be gnu since the whole conda ecosystem is built around gnu and for all Python etc installations you need a working glibc on your system. (I would still download musl for all though)

**Comment by prsabahrami** [comment]
@pavelzw I have added a `header.ps1` which adds support for windows now. 
Let me know what do you think!

**Comment by prsabahrami** [comment]
@pavelzw Any update on this?

File: src/pack.rs [hunk 2]
```diff
@@ -274,6 +294,116 @@ async fn archive_directory(input_dir: &Path, archive_target: &Path) -> Result<()
...
+    let archive_base64 = STANDARD.encode(&compressor);
+    final_executable
+        .write_all(archive_base64.as_bytes())
+        .await?;
+
+    final_executable.write_all(b"\n").await?;
+    if platform.is_windows() {
+        final_executable.write_all(b"__END_ARCHIVE__\n").await?;
+    } else {
+        final_executable.write_all(b"@@END_ARCHIVE@@\n").await?;
```
[the remark this answers is not in the document]
Comment (pavelzw) [reply src/pack.rs:395]: do you know whether conda-pack does it the same way?

File: src/unpack.rs [hunk 3]
```diff
@@ -199,21 +201,22 @@ async fn create_prefix(channel_dir: &Path, target_prefix: &Path, cache_dir: &Pat
...
                         )
-                        .await
-                        .map_err(|e| {
-                            anyhow!(
-                                "could not extract \"{}\": {}",
-                                repodata_record.as_ref().name.as_source(),
-                                e
-                            )
-                        })?;
+                    })?;
```
Comment (pavelzw) [inline src/unpack.rs:219]: is this change necessary / why did you adjust this?

File: tests/integration_test.rs [hunk 4]
```diff
@@ -268,3 +274,82 @@ async fn test_pypi_ignore(
     let pack_result = pixi_pack::pack(pack_options).await;
     assert_eq!(pack_result.is_err(), should_fail);
 }
+#[rstest]
+#[tokio::test]
+async fn test_run_packed_executable(options: Options, required_fs_objects: Vec<&'static str>) {
+    let temp_dir = tempfile::tempdir().expect("Couldn't create a temp dir for tests");
```
Comment (pavelzw) [inline tests/integration_test.rs:280]:
```suggestion
    let temp_dir = tempfile::tempdir().unwrap();
```

unwrap is fine imo as this is not user facing

File: tests/integration_test.rs [hunk 5]
```diff
@@ -268,3 +274,82 @@ async fn test_pypi_ignore(
...
+
+    #[cfg(target_os = "windows")]
+    let activate_file = options.output_dir.path().join("activate.bat");
+    #[cfg(any(target_os = "linux", target_os = "macos"))]
+    let activate_file = options.output_dir.path().join("activate.sh");
+
+    assert!(
+        activate_file.exists(),
+        "Activation script not found after extraction"
+    );
```
Comment (pavelzw) [inline tests/integration_test.rs:344]:
```suggestion
    #[cfg(target_os = "windows")]
    let activation_script = options.output_dir.path().join("activate.bat");
    #[cfg(any(target_os = "linux", target_os = "macos"))]
    let activation_script = options.output_dir.path().join("activate.sh");

    assert!(
        activation_script.exists(),
        "Activation script not found after extraction"
    );
```

File: tests/integration_test.rs [hunk 6]
```diff
@@ -268,3 +274,82 @@ async fn test_pypi_ignore(
     let pack_result = pixi_pack::pack(pack_options).await;
     assert_eq!(pack_result.is_err(), should_fail);
 }
+#[rstest]
+#[tokio::test]
+async fn test_run_packed_executable(options: Options, required_fs_objects: Vec<&'static str>) {
```
Comment (pavelzw) [inline tests/integration_test.rs:279]: i'm wondering whether we can automatically test cross-platform packs in ci 🤔 although i can't think of a pretty way to do so... we should still at least test once that it works

File: src/header.ps1
Comment (pavelzw) [inline src/header.ps1:1]: @delsner please also take a look at this as you know powershell more than me probably

File: src/header.ps1 [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+$VERBOSE = $false
+$QUIET = $false
+$UNPACK_SHELL = ""
+
+$USAGE = @"
+usage: $($MyInvocation.MyCommand.Name) [options]
+
+Unpacks an environment packed using pixi-pack
+
+-f, --force                 No error if environment already exists
```
Comment (pavelzw) [inline src/header.ps1:20]: see my comments in `header.sh`

File: src/header.ps1 [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+# Check if verbose and quiet are both set
+if ($VERBOSE -and $QUIET) {
+    Write-Error "ERROR: Verbose and quiet options cannot be used together."
+    exit 1
+}
+
+# Step 1: Extract the archive and pixi-pack executable, and decode them
+$scriptContent = Get-Content -Raw -Path $MyInvocation.MyCommand.Path
+$lines = $scriptContent -split "`r?`n"
```
Comment (pavelzw) [inline src/header.ps1:55]: is this `\r\n`? 😅

File: src/header.ps1 [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    if ($lines[$i] -like "*__END_HEADER__*") {
+        $headerLine = $i + 1
+    }
+    if ($lines[$i] -like "*__END_ARCHIVE__*") {
+        $archiveLine = $i + 1
+    }
+}
+
+if (-not $headerLine -or -not $archiveLine) {
+    Write-Error "Markers __END_HEADER__ or __END_ARCHIVE__ not found."
```
Comment (pavelzw) [inline src/header.ps1:71]: i don't think we have a similar test in `header.sh`. do you think we can add it there as well?

File: src/header.ps1 [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+$archiveContent = $lines[($headerLine)..($archiveLine - 2)] -join ""
+$archiveContent = $archiveContent.Trim()
+
+# Decode Base64 content into tar file
+try {
+    $decodedArchive = [System.Convert]::FromBase64String($archiveContent)
+    $archivePath = "$TEMPDIR\archive.tar"
+    [System.IO.File]::WriteAllBytes($archivePath, $decodedArchive)
+} catch {
+    Write-Error "Failed to decode Base64 archive content: $_"
```
Comment (pavelzw) [inline src/header.ps1:85]: What does `$_` do?

File: src/header.ps1 [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+# Add shell flag if provided
+if ($UNPACK_SHELL) {
+    $arguments += "--shell"
+    $arguments += $UNPACK_SHELL
+}
+
+# Finally, add the path to the archive
+$arguments += $archivePath
+
+& $pixiPackPath @arguments
```
Comment (pavelzw) [inline src/header.ps1:128]: Is `&` the powershell equivalent of `exec`?

File: tests/integration_test.rs
Comment (pavelzw) [inline tests/integration_test.rs:1]: we added tests for reproducible pixi-packs in #43

https://github.com/Quantco/pixi-pack/blob/c84abb886cb30de003fb2fbc5882a12e2fb9aa56/tests/integration_test.rs#L284

could you extend that test with `create_executable` as well?

Comment (prsabahrami) [reply tests/integration_test.rs:1]: Absolutely. 
I'll try to respond to all of the previous messages this weekend as well. 
I have been extremely busy this week.

File: src/header.sh [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    rm -rf "$PREFIX"
+fi
+
+archive_begin=$(($(grep -anm 1 "^@@END_HEADER@@" "$0" | sed 's/:.*//') + 1))
+archive_end=$(($(grep -anm 1 "^@@END_ARCHIVE@@" "$0" | sed 's/:.*//') - 1))
+
+echo "Unpacking payload ..."
+tail -n +$archive_begin "$0" | head -n $(($archive_end - $archive_begin + 1)) | base64 -d > "$TEMPDIR/archive.tar"
+
+pixi_pack_start=$(($archive_end + 2))
```
[the remark this answers is not in the document]
Comment (prsabahrami) [reply src/header.sh:104]: It's because `archive_end` is actually the last line of archive binary (i.e. the line of `END_ARCHIVE` - 1). In the case of this code where we're encoding into base64, each binary is actually one single line and hence we can actually only find the line number of `@@END_HEADER@@` or (`__END_HEADER__`) and find the others through that line but I thought it's best to keep both of the variables as it's simply more reliable.

File: src/pack.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (prsabahrami) [reply src/pack.rs:452]: I don't think they are exactly comments. I believe they simply have no effect. I got this from [conda-constructor](https://github.com/conda/constructor/blob/d4a35e96890a811688e218c0fa0a683915e8a40b/constructor/header.sh#L724). Please let me know if you have other way of indicating the end of the header file.

File: src/header.ps1 (hunk 8 above)
Comment (prsabahrami) [reply src/header.ps1:51]: I believe so. 😅

File: src/header.ps1 (hunk 10 above)
Comment (prsabahrami) [reply src/header.ps1:81]: IINW It represents the error message in the try-catch block.

File: src/header.ps1 (hunk 11 above)
Comment (prsabahrami) [reply src/header.ps1:124]: In this case yes. I think there are some differences but generally it is the equivalent.

File: src/header.sh [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      exit 1
+      ;;
+    : )
+      echo "Option -$OPTARG requires an argument" >&2
+      echo "$USAGE" >&2
+      exit 1
+      ;;
+  esac
+done
+shift $((OPTIND -1))
```
[the remark this answers is not in the document]
Comment (prsabahrami) [reply src/header.sh]: Removed it since I replaced getopts.

File: tests/integration_test.rs (hunk 6 above)
Comment (prsabahrami) [reply tests/integration_test.rs:405]: That would be amazing.

File: src/header.sh [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    echo "If you want to update an existing environment, use the -f option." >&2
+    exit 1
+fi
+
+if [ "$FORCE" = "1" ] && [ -n "$PREFIX" ] && [ -e "$PREFIX" ]; then
+    rm -rf "$PREFIX"
+fi
+
+archive_begin=$(($(grep -anm 1 "^@@END_HEADER@@" "$0" | sed 's/:.*//') + 1))
+archive_end=$(($(grep -anm 1 "^@@END_ARCHIVE@@" "$0" | sed 's/:.*//') - 1))
```
[the remark this answers is not in the document]
Comment (prsabahrami) [reply src/header.sh]: This can be replaced with awk (I already pushed replacement) but awk on linux and macos are still not the same thing. I tested it on both platforms and it seems to be working fine but I guess if we want to make sure they are exactly the same we'll have to use gawk on linux and nawk on macos. I would say this won't turn out very beautiful but is definitely possible.

File: src/header.sh [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+
+-f, --force                 No error if environment already exists
+-h, --help                  Print this help message and exit
+-o, --output-directory <DIR> Where to unpack the environment
+-s, --shell <SHELL>         Sets the shell [options: bash, zsh, xonsh, cmd, powershell, fish, nushell]
+-v, --verbose               Increase logging verbosity
+-q, --quiet                 Decrease logging verbosity
+"
+# Parse command-line options
+while getopts ":hfvo:s:q" opt; do
```
[the remark this answers is not in the document]
Comment (prsabahrami) [reply src/header.sh]: What I have implemented right now only supports separate args passed onto the cli but I believe since both of the `-o` and `-s` take an argument afterwards it should be fine.

File: tests/integration_test.rs [hunk 16]
```diff
@@ -362,3 +398,85 @@ async fn test_custom_env_name(options: Options) {
...
+    pack_options.create_executable = true;
+
+    #[cfg(target_os = "windows")]
+    {
+        pack_options.output_file = temp_dir.path().join("environment.ps1");
+    }
+    #[cfg(any(target_os = "linux", target_os = "macos"))]
+    {
+        pack_options.output_file = temp_dir.path().join("environment.sh");
+    }
```
Comment (pavelzw) [inline tests/integration_test.rs:416]:
```suggestion
    #[cfg(target_os = "windows")]
    pack_options.output_file = temp_dir.path().join("environment.ps1");
    #[cfg(not(target_os = "windows"))]
    pack_options.output_file = temp_dir.path().join("environment.sh");
```

File: tests/integration_test.rs [hunk 17]
```diff
@@ -362,3 +398,85 @@ async fn test_custom_env_name(options: Options) {
...
+    let pack_result = pixi_pack::pack(pack_options).await;
+    assert!(pack_result.is_ok(), "{:?}", pack_result);
+
+    assert!(
+        pack_file.exists(),
+        "Pack file does not exist at {:?}",
+        pack_file
+    );
+
+    #[cfg(any(target_os = "linux", target_os = "macos"))]
```
Comment (pavelzw) [inline tests/integration_test.rs:429]:
```suggestion
    #[cfg(not(target_os = "windows"))]
```

File: tests/integration_test.rs [hunk 18]
```diff
@@ -362,3 +398,85 @@ async fn test_custom_env_name(options: Options) {
...
+        pack_options.output_file = temp_dir.path().join("environment.sh");
+    }
+
+    let pack_file = pack_options.output_file.clone();
+
+    let pack_result = pixi_pack::pack(pack_options).await;
+    assert!(pack_result.is_ok(), "{:?}", pack_result);
+
+    assert!(
+        pack_file.exists(),
```
Comment (pavelzw) [inline tests/integration_test.rs:424]: let's also assert here that the executable bits are set

File: tests/integration_test.rs [hunk 19]
```diff
@@ -362,3 +398,85 @@ async fn test_custom_env_name(options: Options) {
...
+        assert_eq!(pack_file.extension().unwrap(), "ps1");
+        let output = Command::new("powershell")
+            .arg("-File")
+            .arg(&pack_file)
+            .arg("-o")
+            .arg(options.output_dir.path())
+            .output()
+            .expect("Failed to execute packed file for extraction");
+        assert!(output.status.success(), "Packed file execution failed");
+    }
```
Comment (pavelzw) [inline tests/integration_test.rs:455]: let's switch the order here, in the other examples you always had windows before unix

File: tests/integration_test.rs [hunk 20]
```diff
@@ -362,3 +398,85 @@ async fn test_custom_env_name(options: Options) {
...
+
+    let env_dir = options.output_dir.path().join("env");
+    assert!(
+        env_dir.exists(),
+        "Environment directory not found after extraction"
+    );
+
+    #[cfg(target_os = "windows")]
+    let activation_script = options.output_dir.path().join("activate.bat");
+    #[cfg(any(target_os = "linux", target_os = "macos"))]
```
Comment (pavelzw) [inline tests/integration_test.rs:465]:
```suggestion
    #[cfg(not(target_os = "windows"))]
```

File: tests/integration_test.rs [hunk 21]
```diff
@@ -362,3 +398,85 @@ async fn test_custom_env_name(options: Options) {
...
+    {
+        assert_eq!(pack_file.extension().unwrap(), "sh");
+        let output = Command::new("bash")
+            .arg(&pack_file)
+            .arg("-o")
+            .arg(options.output_dir.path())
+            .output()
+            .expect("Failed to execute packed file for extraction");
+        assert!(
```
Comment (pavelzw) [inline tests/integration_test.rs:438]: let's move this assert out of the `#[cfg...` block

**Review by pavelzw — COMMENTED** [review]
awesome! getting there :)

File: tests/integration_test.rs [hunk 22]
```diff
@@ -310,6 +316,36 @@ async fn test_reproducible_shasum(options: Options) {
     let pack_result = pixi_pack::pack(pack_options.clone()).await;
     assert!(pack_result.is_ok(), "{:?}", pack_result);
 
+    // Second pack.
+    pack_options.output_file = output_file2.clone();
+    let pack_result = pixi_pack::pack(pack_options.clone()).await;
+    assert!(pack_result.is_ok(), "{:?}", pack_result);
+
+    assert_eq!(
```
Comment (pavelzw) [inline tests/integration_test.rs:324]: i found a nicer way to test reproducibility using insta, see #75, so this test needs to be adjusted when merging main 😓 sorry for the extra effort

File: src/main.rs [hunk 23]
```diff
@@ -87,6 +91,18 @@ enum Commands {
     },
 }
 
+fn default_output_file(create_executable: bool) -> PathBuf {
+    if create_executable {
+        if cfg!(target_os = "windows") {
```
[the remark this answers is not in the document]
Comment (prsabahrami) [reply src/main.rs:96]: Oh yeahh. I don't know why I did that. My bad.

File: tests/integration_test.rs [hunk 24]
```diff
@@ -30,9 +30,14 @@ fn options(
     #[default(Some(ShellEnum::Bash(Bash)))] shell: Option<ShellEnum>,
     #[default(false)] ignore_pypi_errors: bool,
     #[default("env")] env_name: String,
+    #[default(false)] create_executable: bool,
 ) -> Options {
     let output_dir = tempdir().expect("Couldn't create a temp dir for tests");
-    let pack_file = output_dir.path().join("environment.tar");
+    let pack_file = if create_executable {
+        output_dir.path().join("environment.sh")
```
[the remark this answers is not in the document]
Comment (prsabahrami) [reply tests/integration_test.rs:37]: ah yes I forgot to update this from when windows support was not added.

File: src/header.ps1 [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+
+-h, --help                  Print this help message and exit
+-o, --output-directory <DIR> Where to unpack the environment
+-s, --shell <SHELL>         Sets the shell [options: bash, zsh, xonsh, cmd, powershell, fish, nushell]
+-v, --verbose               Increase logging verbosity
+-q, --quiet                 Decrease logging verbosity
+"@
+
+# Parse command-line arguments
+$InputArgs = $MyInvocation.UnboundArguments
```
[the remark this answers is not in the document]
Comment (prsabahrami) [reply src/header.ps1]: Agreed. I tried this on macos and works beautifully.

File: src/pack.rs [hunk 26]
```diff
@@ -328,6 +343,130 @@ async fn archive_directory(input_dir: &Path, archive_target: &Path) -> Result<()
...
+    }
+
+    // Encode the executable to base64
+    let executable_base64 = STANDARD.encode(&executable_bytes);
+    final_executable
+        .write_all(executable_base64.as_bytes())
+        .await?;
+
+    // Make the script executable
```
Comment (pavelzw) [inline src/pack.rs:462]: let's add a comment that this won't be executed when cross-packing but is not possible otherwise since the FS most likely doesn't support it

File: src/header.sh [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+archive_end=$((archive_end - 1))
+pixi_pack_start=$(($archive_end + 2))
+
+echo $(tail -n +$archive_begin "$0" | head -n $(($archive_end - $archive_begin + 1))) | base64 -d > "$TEMPDIR/archive.tar"
+echo $(tail -n +$pixi_pack_start "$0") | base64 -d > "$TEMPDIR/pixi-pack"
+
+chmod +x "$TEMPDIR/pixi-pack"
+
+"$TEMPDIR/pixi-pack" unpack $@ "$TEMPDIR/archive.tar"
+
```
Comment (pavelzw) [inline src/header.sh:46]: let's remove the tempdir at the end

File: src/header.ps1 [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+
+# Build the command with flags
+$arguments = @("unpack")
+$arguments += $args | Join-String -Separator ' '
+
+# Add the path to the archive
+$arguments += $archivePath
+
+& $pixiPackPath @arguments
+
```
Comment (pavelzw) [inline src/header.ps1:89]: let's remove the tempdir at the end

**Review by pavelzw — APPROVED** [review]
the only thing missing now is adding the feature to the README, i can take care of this :)

File: src/header.sh [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+  echo "ERROR: Markers @@END_HEADER@@ or @@END_ARCHIVE@@ not found." >&2
+  exit 1
+fi
+
+archive_begin=$((archive_begin + 2))
+archive_end=$((archive_end - 1))
+pixi_pack_start=$(($archive_end + 2))
+
+echo $(tail -n +$archive_begin "$0" | head -n $(($archive_end - $archive_begin + 1))) | base64 -d > "$TEMPDIR/archive.tar"
+echo $(tail -n +$pixi_pack_start "$0") | base64 -d > "$TEMPDIR/pixi-pack"
```
Comment (pavelzw) [inline src/header.sh:41]: this is not the most efficient way of doing this...

```
❯ time tail -n +64 environment.sh | head -n 1 | base64 -d > /tmp/archive.tar

________________________________________________________
Executed in   22.93 secs    fish           external
   usr time   20.63 secs    0.13 millis   20.63 secs
   sys time    1.29 secs   12.02 millis    1.28 secs
```

takes 22s on my macbook

while

```
❯ time sed -n "64p" environment.sh | base64 -d > /tmp/archive.tar

________________________________________________________
Executed in    7.45 secs    fish           external
   usr time    6.10 secs    0.11 millis    6.10 secs
   sys time    0.86 secs    5.41 millis    0.86 secs
```

takes only 7s.

does this sed command work with both macos and linux?

Comment (prsabahrami) [reply src/header.sh:41]: I tried it on BSD and read through gnu's doc. It should be fine.

Comment (pavelzw) [reply src/header.sh]: this only works though, if the archive base64 is in a singe line (which it is, right?)
(we should assert this in the shell script somewhere

Comment (prsabahrami) [reply src/header.sh]: Yes in our case it is. I'll do that. But also why only one line? I think this works for multiple lines as well, no?
```sh
sed -n "$i,$jp" 
```

File: src/header.sh
[the remark this answers is not in the document]
Comment (pavelzw) [reply src/header.sh:1]: i pushed to your pr @prsabahrami [`35df7c0` (#31)](https://github.com/Quantco/pixi-pack/pull/31/commits/35df7c0b00957ff187233b40a662629d79019e4d)

File: tests/integration_test.rs (hunk 6 above)
Comment (pavelzw) [reply tests/integration_test.rs:429]: this is not necessary anymore since we ensure the shasum is reproducible between platforms

**Review by pavelzw — APPROVED** [review]
Really nice @prsabahrami! Thanks for persevering, looking forward to using this feature 🤩 

note: tests are failing because it tries to download pixi-pack 0.3.0 in the create-executable tests which is not yet released
